Tactics/Reflect/FetchAndDecode.lean

```diff
@@ -0,0 +1,58 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Alex Keizer
+-/
+import Arm.State
+import Tactics.Common
+import Tactics.Reflect.ProgramInfo
+
+open Lean Meta
+open Elab.Tactic Elab.Term
+
+initialize
+  Lean.registerTraceClass `Sym.reduceFetchInst
+
+def reduceFetchInst? (addr : Expr) (s : Expr) :
+    MetaM (BitVec 32 × Expr) := do
+  let e := mkApp2 (mkConst ``fetch_inst) addr s
+
+  let addr ← reflectBitVecLiteral 64 addr
+  let ⟨programHyp, program⟩ ← findProgramHyp s
+  let programInfo ← try ProgramInfo.lookupOrGenerate program catch err =>
+    throwErrorAt
+      err.getRef
+      "Could not generate ProgramInfo for {program}:\n\n{err.toMessageData}"
+
+  let some rawInst := programInfo.getRawInstrAt? addr
+    | throwError "No instruction found at address {addr}"
+
+  trace[Sym.reduceFetchInst] "{Lean.checkEmoji} reduced to: {rawInst}"
+
+  -- Now, construct the proof
+  let eqType ← mkEq e (toExpr (some rawInst))
+  trace[Sym.reduceFetchInst] "⚙️ deriving a proof of:\n\t{eqType}"
+
+  let h_program := mkIdent programHyp.userName
+  let eqProofStx ← `(term| by
+    unfold fetch_inst; rw [$h_program:ident]; decide
```

Comment:
Will `decide` here cause maxRecDepth to be reached for really large programs?

Reply:
I've simplified this, by extracting out a `fetch_inst_eq_of_prgram_eq_of_map_find` lemma.
The simproc now directly builds an application of that theorem, so that part is instant.

The only risk is that the proof still uses `rfl`, so for truly ginormous programs, kernel reduction might slow down. However, tests for SHA512 are instantaneous ther; it should not be a problem.

Reply:
@alexkeizer I guess we can have an option that (defaults to) `native_decide`, and uses `decide` if the option (name subject to bikeshed: `minimalTCB := true`) is enabled?

Reply:
Post simplification we don't use `decide` at all, only `rfl`. Since it's still instantaneous for SHA512, I don't think we have to worry about this at all

Reply:
Thanks, @alexkeizer. The addition of `fetch_inst_eq_of_prgram_eq_of_map_find` addresses my concern.

BTW, there are other large programs we can test this on. E.g., see https://github.com/leanprover/LNSym/blob/main/Tests/AES-GCM/AESGCMEncKernelProgram.lean.